General cleanup of FilterOperations

weinig opened the pull request:
#### 265f430df9e58c0d9c2ebcea32a2d4da4f047d59
<pre>
General cleanup of FilterOperations
<a href="https://bugs.webkit.org/show_bug.cgi?id=274781">https://bugs.webkit.org/show_bug.cgi?id=274781</a>

Reviewed by Darin Adler.

Makes some small improvements to the FilterOperations type:

- Use Ref instead of RefPtr in operations vector.
- Made immutable. All construction already happened in one shot in the builder
  or blend functions, or was a full replacement, so this just meant removing
  non-const accessors and removing the clear() function.
- Expose iterators directly to avoid clients accessing the underlying vector.

The fallout of using Ref for the operations vector led to some fallout requiring
more use of Ref or reference types to indicate non-null state.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForFilter):
    - Use isEmpty() from FilterOperations directly, and update for Ref values
      during iteration.

* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::AcceleratedEffectValues):
(WebCore::AcceleratedEffectValues::clone const):
    - Use clone() helper and copy construction.

* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::validateFilterOperations):
     - Use isEmpty() from FilterOperations directly.

* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::FilterAnimationValue::FilterAnimationValue):
    - Use clone() helper.

(WebCore::GraphicsLayer::clearFilters):
(WebCore::GraphicsLayer::clearBackdropFilters):
    - Re-initialize the whole type rather than clearing internal state.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createFilterAnimationsFromKeyframes):
(WebCore::GraphicsLayerCA::setFilterAnimationEndpoints):
(WebCore::GraphicsLayerCA::setFilterAnimationKeyframes):
    - Update to account for Ref FilterOperations.

* Source/WebCore/platform/graphics/ca/PlatformCAAnimation.h:
* Source/WebCore/platform/graphics/ca/PlatformCAFilters.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAAnimationCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAAnimationCocoa.mm:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
    - Update to use FilterOperation&amp; and Ref&lt;FilterOperation&gt; parameters. All
      implementations already assumed the values were non-null, so now the types
      also know it.

* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
(WebCore::FilterOperations::FilterOperations):
(WebCore::FilterOperations::operator== const):
(WebCore::FilterOperations::operationsMatch const):
    - Update to account for Ref values in m_operations.

(WebCore::FilterOperations::hasReferenceFilter const):
    - Use new hasFilterOfType&lt;T&gt; helper.

(WebCore::FilterOperations::outsets const):
    - Update to account for Ref values in m_operations.

(WebCore::FilterOperations::hasFilterThatAffectsOpacity const):
(WebCore::FilterOperations::hasFilterThatMovesPixels const):
(WebCore::FilterOperations::hasFilterThatShouldBeRestrictedBySecurityOrigin const):
    - Use WTF::anyOf rather than explicit loop.

(WebCore::FilterOperations::blend const):
    - Use explicit Vector&lt;Ref&lt;FilterOperation&gt;&gt; for constructing blended results,
      allowing for the use of reserveInitialCapacity optimization.

(WebCore::operator&lt;&lt;):
    - Remove now unnecessary null-check and use existing container serializer
      instead of re-implementing.

* Source/WebCore/platform/graphics/filters/FilterOperations.h:
(WebCore::FilterOperations::FilterOperations):
    - Add copy and move constructors/assignment operators.

(WebCore::FilterOperations::clone const):
    - Add helper to create a clone() of each operation.

(WebCore::FilterOperations::operations const):
    - Remove non-const overloads.

(WebCore::FilterOperations::at const):
    - Update to account for Ref in m_operations.

(WebCore::FilterOperations::hasFilterOfType const):
    - Helper which can check if any operation is a particular type.

(WebCore::FilterOperations::clear): Deleted.
(WebCore::FilterOperations::operations): Deleted.
(WebCore::FilterOperations::setOperations): Deleted.
    - Remove member functions that mutated the underlying vector.

* Source/WebCore/style/FilterOperationsBuilder.cpp:
(WebCore::Style::createFilterOperations):
    - Use explicit Vector&lt;Ref&lt;FilterOperation&gt;&gt; for constructing results
      and shrinkToFit at the end.

* Source/WebCore/rendering/CSSFilter.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
    - Update to account for Ref in m_operations.

* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasAppleColorFilter const):
(WebCore::RenderStyle::hasFilter const):
(WebCore::RenderStyle::hasBackdropFilter const):
    - Use isEmpty() directly on FilterOperations.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
    - Update to account for Ref in m_operations.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::updatePageBackgroundLayers):
    - Construct FilterOperations directly rather than using temporary vector.

* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemote.mm:
(WebKit::PlatformCAAnimationRemote::setFromValue):
(WebKit::PlatformCAAnimationRemote::setToValue):
(WebKit::PlatformCAAnimationRemote::setValues):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemoteProperties.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemoteProperties.serialization.in:
    - Update for use of Ref and reference types for FilterOperation.

Canonical link: <a href="https://commits.webkit.org/279490@main">https://commits.webkit.org/279490@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b6bc67850b5d026bb15f640f400d671fc77e3b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wincairo-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11686 "Built successfully and passed tests") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-10-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/platform/graphics/filters/FilterOperations.h:
```diff
@@ -37,20 +38,37 @@ struct BlendingContext;
+    FilterOperations clone() const
+    {
+        return FilterOperations { m_operations.map([](const auto& op) { return op->clone(); }) };
+    }
```
Having both a `copy` and `clone` is confusing. 
Do we really need both ? If yes, should we consider other names ? shallowCopy/deepCopy ?

weinig replied:
We can just remove copy() and make FilterOperations do the right thing under copy construction.

darinadler commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
```diff
@@ -353,16 +356,18 @@ static unsigned keyValueCountForFilter(const FilterOperation& filterOperation)
-    
-    if (auto* defaultOperation = dynamicDowncast<DefaultFilterOperation>(*operation)) {
+    const FilterOperation* operation = &filterOperation;
+    FilterOperation::Type type = filterOperation.type();
```
I might just write `auto` here

darinadler commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
```diff
@@ -353,16 +356,18 @@ static unsigned keyValueCountForFilter(const FilterOperation& filterOperation)
-    RetainPtr<id> value;
-    
-    if (auto* defaultOperation = dynamicDowncast<DefaultFilterOperation>(*operation)) {
+    const FilterOperation* operation = &filterOperation;
```
I might just write `auto*` or `const auto*` here.

darinadler commented on Source/WebCore/platform/graphics/filters/FilterOperations.h:
```diff
@@ -37,20 +38,37 @@ struct BlendingContext;
+    FilterOperations(const FilterOperations&) = default;
+    FilterOperations(FilterOperations&&) = default;
+    FilterOperations& operator=(const FilterOperations&) = default;
+    FilterOperations& operator=(FilterOperations&&) = default;
```
Why do we have to write this out explicitly?

weinig replied:
No. Not sure why I thought I did.

darinadler commented on Source/WebCore/style/FilterOperationsBuilder.cpp:
```diff
@@ -192,7 +192,9 @@ std::optional<FilterOperations> createFilterOperations(const Document& document,
     }
 
-    return operations;
+    operations.shrinkToFit();
```
Somehow I thought that move already did this?

weinig replied:
Vector move is just really a pointer swap, no allocation changes.

Commit:
General cleanup of FilterOperations
https://bugs.webkit.org/show_bug.cgi?id=274781

Reviewed by Darin Adler.

Makes some small improvements to the FilterOperations type:

- Use Ref instead of RefPtr in operations vector.
- Made immutable. All construction already happened in one shot in the builder
  or blend functions, or was a full replacement, so this just meant removing
  non-const accessors and removing the clear() function.
- Expose iterators directly to avoid clients accessing the underlying vector.

The fallout of using Ref for the operations vector led to some fallout requiring
more use of Ref or reference types to indicate non-null state.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForFilter):
    - Use isEmpty() from FilterOperations directly, and update for Ref values
      during iteration.

* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::AcceleratedEffectValues):
(WebCore::AcceleratedEffectValues::clone const):
    - Use clone() helper and copy construction.

* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::validateFilterOperations):
     - Use isEmpty() from FilterOperations directly.

* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::FilterAnimationValue::FilterAnimationValue):
    - Use clone() helper.

(WebCore::GraphicsLayer::clearFilters):
(WebCore::GraphicsLayer::clearBackdropFilters):
    - Re-initialize the whole type rather than clearing internal state.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createFilterAnimationsFromKeyframes):
(WebCore::GraphicsLayerCA::setFilterAnimationEndpoints):
(WebCore::GraphicsLayerCA::setFilterAnimationKeyframes):
    - Update to account for Ref FilterOperations.

* Source/WebCore/platform/graphics/ca/PlatformCAAnimation.h:
* Source/WebCore/platform/graphics/ca/PlatformCAFilters.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAAnimationCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAAnimationCocoa.mm:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
    - Update to use FilterOperation& and Ref<FilterOperation> parameters. All
      implementations already assumed the values were non-null, so now the types
      also know it.

* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
(WebCore::FilterOperations::FilterOperations):
(WebCore::FilterOperations::operator== const):
(WebCore::FilterOperations::operationsMatch const):
    - Update to account for Ref values in m_operations.

(WebCore::FilterOperations::hasReferenceFilter const):
    - Use new hasFilterOfType<T> helper.

(WebCore::FilterOperations::outsets const):
    - Update to account for Ref values in m_operations.

(WebCore::FilterOperations::hasFilterThatAffectsOpacity const):
(WebCore::FilterOperations::hasFilterThatMovesPixels const):
(WebCore::FilterOperations::hasFilterThatShouldBeRestrictedBySecurityOrigin const):
    - Use WTF::anyOf rather than explicit loop.

(WebCore::FilterOperations::blend const):
    - Use explicit Vector<Ref<FilterOperation>> for constructing blended results,
      allowing for the use of reserveInitialCapacity optimization.

(WebCore::operator<<):
    - Remove now unnecessary null-check and use existing container serializer
      instead of re-implementing.

* Source/WebCore/platform/graphics/filters/FilterOperations.h:
(WebCore::FilterOperations::FilterOperations):
    - Add copy and move constructors/assignment operators.

(WebCore::FilterOperations::clone const):
    - Add helper to create a clone() of each operation.

(WebCore::FilterOperations::operations const):
    - Remove non-const overloads.

(WebCore::FilterOperations::at const):
    - Update to account for Ref in m_operations.

(WebCore::FilterOperations::hasFilterOfType const):
    - Helper which can check if any operation is a particular type.

(WebCore::FilterOperations::clear): Deleted.
(WebCore::FilterOperations::operations): Deleted.
(WebCore::FilterOperations::setOperations): Deleted.
    - Remove member functions that mutated the underlying vector.

* Source/WebCore/style/FilterOperationsBuilder.cpp:
(WebCore::Style::createFilterOperations):
    - Use explicit Vector<Ref<FilterOperation>> for constructing results
      and shrinkToFit at the end.

* Source/WebCore/rendering/CSSFilter.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
    - Update to account for Ref in m_operations.

* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasAppleColorFilter const):
(WebCore::RenderStyle::hasFilter const):
(WebCore::RenderStyle::hasBackdropFilter const):
    - Use isEmpty() directly on FilterOperations.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
    - Update to account for Ref in m_operations.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::updatePageBackgroundLayers):
    - Construct FilterOperations directly rather than using temporary vector.

* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemote.mm:
(WebKit::PlatformCAAnimationRemote::setFromValue):
(WebKit::PlatformCAAnimationRemote::setToValue):
(WebKit::PlatformCAAnimationRemote::setValues):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemoteProperties.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemoteProperties.serialization.in:
    - Update for use of Ref and reference types for FilterOperation.

Canonical link: https://commits.webkit.org/279490@main
